pilot/pkg/model/gateway.go

```diff
@@ -68,6 +69,19 @@ func MergeGateways(gateways ...Config) *MergedGateway {
 			gatewayNameForServer[s] = gatewayName
 			log.Debugf("MergeGateways: gateway %q processing server %v", gatewayName, s.Hosts)
 			protocol := ParseProtocol(s.Port.Protocol)
+
+			if s.Tls != nil {
+				// Envoy will reject config that has multiple filter chain matches with the same matching rules
+				// To avoid this, we need to make sure we don't have duplicated hosts, which will become
+				// SNI filter chain matches
+				if tlsHostsByPort[s.Port.Number] == nil {
+					tlsHostsByPort[s.Port.Number] = map[string]struct{}{}
+				}
+				if duplicateHosts := checkDuplicates(s.Hosts, tlsHostsByPort[s.Port.Number]); len(duplicateHosts) != 0 {
+					log.Warnf("MergeGateways: skipping server on gateway %s, duplicate host names: %v", gatewayName, duplicateHosts)
+					continue
```

Comment:
Why do you have to reject the entire server? Just remove the host or reject the host.

Reply:
Good idea. One minor concern - do we guarantee (or want to) guarantee atomicity of config? What I mean is it ok to accept part of a config, or should we accept all or nothing?

I don't think this is something we need to guarantee though, so I can change this to just remove the conflicting host

Reply:
thats a fair point. thats what we do for the port names as well. The thing is, the duplicate host stuff worked before (dunno how but somehow envoy didn't barf). 
I guess its okay. Better to reject the entire server than just one host, as the server has certs and other stuff.

Reply:
In 1.1.2 and below it was "working", but Envoy would just pick an arbitrary one (I think the last one). However it could crash envoy in some instances - see https://github.com/istio/istio/issues/13735.

In 1.1.4 we picked up this PR envoyproxy/envoy#6022 which always blocks it